### PR TITLE
Forking the posix poll based poller for iomgr event engine

### PR DESCRIPTION
**Pull request by Vignesh2208**
Fork of the posix poll based poller implementation for iomgr event engine

### CODE REVIEW

File: test/core/event_engine/iomgr_event_engine/event_poller_posix_test.cc [hunk 1]
```diff
@@ -479,16 +480,20 @@ int main(int argc, char** argv) {
...
+  int result = 0;
   EXPECT_NE(engine, nullptr);
   grpc_event_engine::iomgr_engine::TestScheduler scheduler(engine.get());
-  g_event_poller =
-      grpc_event_engine::iomgr_engine::GetDefaultPoller(&scheduler);
-  if (g_event_poller == nullptr) {
-    // Poller is not supported on this system.
-    return 0;
+  // Try both pollers (epoll1 and poll) one after the other.
+  g_event_poller = grpc_event_engine::iomgr_engine::GetEpoll1Poller(&scheduler);
```
Comment (ctiller) [inline test/core/event_engine/iomgr_event_engine/event_poller_posix_test.cc:487]: Let's do a test fixture, and parameterize the test based on poller implementation?

File: src/core/lib/event_engine/iomgr_engine/event_poller_posix_default.cc [hunk 2]
```diff
@@ -15,13 +15,18 @@
...
 
 #include "src/core/lib/event_engine/iomgr_engine/ev_epoll1_linux.h"
+#include "src/core/lib/event_engine/iomgr_engine/ev_poll_posix.h"
 #include "src/core/lib/event_engine/iomgr_engine/event_poller.h"
 
 namespace grpc_event_engine {
 namespace iomgr_engine {
 
 EventPoller* GetDefaultPoller(Scheduler* scheduler) {
   EventPoller* poller = GetEpoll1Poller(scheduler);
```
[the remark this answers is not in the document]
Comment (Vignesh2208) [reply src/core/lib/event_engine/iomgr_engine/event_poller_posix_default.cc]: Done. I added a grpc_poll_strategy config flag which can a comma separated list of pollers to try in order of preference. Some more code is copied from iomgr/ev_posix.cc to aid in string parsing.

File: src/core/lib/event_engine/iomgr_engine/ev_poll_posix.cc [hunk 3]
```diff
@@ -0,0 +1,835 @@
...
+  int watch_mask_;
+  absl::Status shutdown_error_;
+  IomgrEngineClosure* on_done_;
+  IomgrEngineClosure* read_closure_;
+  IomgrEngineClosure* write_closure_;
+};
+
+namespace {
+
+bool kPollPollerSupported = false;
```
[the remark this answers is not in the document]
Comment (Vignesh2208) [reply src/core/lib/event_engine/iomgr_engine/ev_poll_posix.cc:221]: I'm not sure what else you had in mind here ? Should I make this static to the PollPoller class ? These variables need to be used by every PollPoller instance

File: src/core/lib/event_engine/iomgr_engine/ev_poll_posix.cc [hunk 4]
```diff
@@ -0,0 +1,835 @@
...
+#include "src/core/lib/event_engine/iomgr_engine/wakeup_fd_posix.h"
+#include "src/core/lib/event_engine/iomgr_engine/wakeup_fd_posix_default.h"
+#include "src/core/lib/gprpp/fork.h"
+#include "src/core/lib/gprpp/time.h"
+
+using ::grpc_event_engine::iomgr_engine::WakeupFd;
+
+#define CLOSURE_NOT_READY ((IomgrEngineClosure*)0)
+#define CLOSURE_READY ((IomgrEngineClosure*)1)
+#define POLLOUT_CHECK (POLLOUT | POLLHUP | POLLERR)
```
[the remark this answers is not in the document]
Comment (Vignesh2208) [reply src/core/lib/event_engine/iomgr_engine/ev_poll_posix.cc]: Done.

File: test/core/event_engine/iomgr_event_engine/event_poller_posix_test.cc (hunk 1 above)
Comment (Vignesh2208) [reply test/core/event_engine/iomgr_event_engine/event_poller_posix_test.cc]: Done. I parametrized instead by poll strategy name to also test the code which creates the default poller to use.

File: BUILD [hunk 5]
```diff
@@ -2508,6 +2508,34 @@ grpc_cc_library(
...
+    ],
+    hdrs = [
+        "src/core/lib/event_engine/iomgr_engine/ev_poll_posix.h",
+    ],
+    external_deps = [
+        "absl/status:statusor",
+        "absl/strings",
+        "absl/synchronization",
+    ],
+    tags = ["grpc-autodeps"],
```
Comment (ctiller) [inline BUILD:2524]: no longer needed!

File: src/core/lib/event_engine/iomgr_engine/event_poller_posix_default.cc [hunk 6]
```diff
@@ -14,14 +14,77 @@
...
+
+GPR_GLOBAL_CONFIG_DECLARE_STRING(grpc_poll_strategy);
 
 namespace grpc_event_engine {
 namespace iomgr_engine {
 
+namespace {
+
+// Adapted from src/core/lib/iomgr/ev_posix.cc
+void add(const char* beg, const char* end, char*** ss, size_t* ns) {
```
Comment (ctiller) [inline src/core/lib/event_engine/iomgr_engine/event_poller_posix_default.cc:34]: remove this (see below)

File: src/core/lib/event_engine/iomgr_engine/event_poller_posix_default.cc [hunk 7]
```diff
@@ -14,14 +14,77 @@
...
+  if (c == nullptr) {
+    add(s, s + strlen(s), ss, ns);
+  } else {
+    add(s, c, ss, ns);
+    split(c + 1, ss, ns);
+  }
+}
+
+// Adapted from src/core/lib/iomgr/ev_posix.cc
+bool is(const char* want, const char* have) {
```
Comment (ctiller) [inline src/core/lib/event_engine/iomgr_engine/event_poller_posix_default.cc:60]:
```
bool PollStrategyMatches(absl::string_view strategy, absl::string_view want) {
  return strategy == "all" || strategy == want;
}
```

File: src/core/lib/event_engine/iomgr_engine/event_poller_posix_default.cc [hunk 8]
```diff
@@ -14,14 +14,77 @@
...
+}  // namespace
+
 EventPoller* GetDefaultPoller(Scheduler* scheduler) {
-  EventPoller* poller = GetEpoll1Poller(scheduler);
+  grpc_core::UniquePtr<char> poll_strategy =
+      GPR_GLOBAL_CONFIG_GET(grpc_poll_strategy);
+  EventPoller* poller = nullptr;
+  char** strings = nullptr;
+  size_t nstrings = 0;
+  split(poll_strategy.get(), &strings, &nstrings);
```
[the remark this answers is not in the document]
Comment (Vignesh2208) [reply src/core/lib/event_engine/iomgr_engine/event_poller_posix_default.cc]: Done. Thanks for pointing this out. didn't know about this util

File: src/core/lib/event_engine/iomgr_engine/event_poller_posix_default.cc [hunk 9]
```diff
@@ -14,14 +14,77 @@
...
+  s = static_cast<char*>(gpr_malloc(len + 1));
+  memcpy(s, beg, len);
+  s[len] = 0;
+  *ss = static_cast<char**>(gpr_realloc(*ss, sizeof(char**) * np));
+  (*ss)[n] = s;
+  *ns = np;
+}
+
+// Adapted from src/core/lib/iomgr/ev_posix.cc
+void split(const char* s, char*** ss, size_t* ns) {
```
[the remark this answers is not in the document]
Comment (Vignesh2208) [reply src/core/lib/event_engine/iomgr_engine/event_poller_posix_default.cc]: Done.

File: src/core/lib/event_engine/iomgr_engine/event_poller_posix_default.cc (hunk 6 above)
Comment (Vignesh2208) [reply src/core/lib/event_engine/iomgr_engine/event_poller_posix_default.cc]: Done.

File: BUILD (hunk 5 above)
Comment (Vignesh2208) [reply BUILD]: Done.

File: src/core/lib/event_engine/iomgr_engine/ev_poll_posix.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (Vignesh2208) [reply src/core/lib/event_engine/iomgr_engine/ev_poll_posix.cc]: I think I see what you mean here. I changed it not to a static variable local to GetPollPoller. I also noticed similar issue in epoll1 poller so I included the change there too.

**Review by ctiller — APPROVED** [review]

**Pull request merged**